### ISSUE TITLE
Binary Checksums and SHA256 signing keys.

### ISSUE THREAD

**Issue by AreDubya** [trezor/trezord-go#42]
I would like to suggest two steps to clear any misconceptions with users about the application's security : 

1. Provide SHA256 checksums for the binary downloads available [on the download page](https://wallet.trezor.io/#/bridge). Actually, these would be nice for any binaries. Now that Windows users can run checksums easily via the Linux Subsystem one file integrity verification method should work for all platforms. 

2. Upgrade [timestamps to SHA256](https://support.comodo.com/index.php?/Knowledgebase/Article/View/1102/38/code-signing-certificates---sha-1-and-sha-256-information). Currently the downloaded Windows .exe's "Digital Signature" tab lists the certificate as being SHA1, when the actual signature for the application is SHA-256. I believe changing the timestamp would get rid of any Windows SHA1 warnings users might see, as the underlying certificate is SHA256. 

![Certificate when checking in "allow this app to make changes" UAC window](https://user-images.githubusercontent.com/520744/37752196-6b0a8a96-2d5c-11e8-89b2-c53815632fd3.jpg)

<img width="305" alt="trezor-bridge-dl-cert" src="https://user-images.githubusercontent.com/520744/37752152-0fc784ea-2d5c-11e8-9492-10645673b361.png">

**Comment by prusnak** [comment]
1. It does not make sense to check against hashes you download over the same channels like binaries. Signed hashes make sense, but then you are reinventing the wheel as binaries are already signed.

2. We'll look into that.

**Comment by prusnak** [comment]
If you want to help with 2:

This line does the signing + timestamping: https://github.com/trezor/trezord-go/blob/master/release/windows/release.sh#L21

If you investigate what to change, we'll apply the change.

**Comment by AreDubya** [comment]
@prusnak,

1. Good point. I think checksums on github is the done thing with downloads elsewhere, but your point about the double signature obviates it I think.
 
2. Changing the timestamp url to include the algorithm as a parameter (http://timestamp.comodoca.com/?td=sha256) should do it. I believe you will want to do that on line 21 and 33. For some signing tools that /td switch would have to be part of the initial command, but after reviewing the docs for osslsigncode I think the url parameter is  all that is needed.

I would submit a PR with those (less than exciting) fixes but I can't test that solution fully. While I was able to get the installers to compile I wasn't able to get xgo to pull the local, edited files; there was a downloading (github) message each time I ran the make. I would experiment with the pathing more, but I don't normally use Go so that could take time. I'm wrapping up construction on a site which will sell Trezors at the moment, so I hope you will excuse my incomplete solution. 

Thanks again for looking into this.

**Comment by karel-3d** [comment]
make downloads the source code, but just the go source I think

Anyway if you want to built locally, just do

`LOCAL_BUILD=1 make all`

as seen in travis

https://github.com/trezor/trezord-go/blob/master/.travis.yml#L23

**Comment by cyphar** [comment]
I'm planning on packaging this for openSUSE. From the perspective of a distribution, the most safe way for a project to provide source code that a distribution can use for building is to provide:

* A deterministic `.tar.gz` of *all* the source code, including any vendored dependencies in Go. This can be generated from `git` pretty easily.
* A detached signature of the above `.tar.gz`.
* A signed set of hashes of the built binaries would also be useful for reproducible build checking, if the project is reproducibly built (which it is since it's a Go project).
* And a link to the PGP key used for the above signing -- preferably in an `gpg-offline` format but that's not required.

I already have a script that does all of this, and is used by both [`runc`](https://github.com/opencontainers/runc/blob/v1.0.0-rc5/script/release.sh) (the runtime that Docker uses) as well as [`umoci`](https://github.com/openSUSE/umoci/blob/v0.4.0/hack/release.sh). This script is written so that you can easily swap out the distribution-specific parts of the build process, but still have the same structure of files to upload.

**Comment by AreDubya** [comment]
@karel-3d 

Sorry for the late reply, I had to meet a few deadlines for the new site. 

I was able to do the local build based on your comment, but I wasn't able to get the `release\windows\release.sh` to sign the package. I think this was due to the key requirements for `osslsigncode` but again, I didn't have time to dig into it fully. I did take the file that was build and sign it with the Windows 10 native version of SignTool, which is what osslsigncode emulates. The following sequence worked for me to generate a test certificate and sign with a sha256 timestamp.

*Create pvk & cer*
`MakeCert /a sha256 /n "CN=AreDubya Test Inc" /r /h 0  /len 4096  /eku "1.3.6.1.5.5.7.3.3,1.3.6.1.4.1.311.10.3.13" /e 04/15/2018 /sv rwKey2.pvk rwKey2.cer`

*Create pfx from generated files*
`Pvk2Pfx /pvk rwKey.pvk /pi <pvk-password> /spc rwKey.cer /pfx rwTrezorTest.pfx /po <pfx-password>`

*Sign file with sha256 timestamp*
`signtool sign /f "C:\Path\To\Certificate\rwTrezorTest.pfx" /fd sha256 /p <pfx-password> /tr "http://timestamp.comodoca.com?td=sha256" /i "AreDubya Test Inc" trezor-bridge-2.0.11-win32-install.exe`

There were a couple of caveats to getting the signature with the SHA256 timestamp : 

- The signing certificate needs to be SHA256.
- Requesting a timestamp from the wrong endpoint will fail. SignTool uses either a `/t <url>` or a `/tr <url>` switch to add a timestamp to the signature. I believe the `/t` (for Authenticode timestamping) is the relevant option, based on the `osslsigncode` repo. If that is the case, then the url with parameter is all that is needed. If not, passing the `/td sha256` switch will be necessary. Details on the two types are [on the Comodo site.](https://support.comodo.com/index.php?/Knowledgebase/Article/View/68/0/time-stamping-server&td=sha256)

 I hope this information is helpful, if only as a research time saver.

The signed file is attached below : 
[signedWithSha256Timestamp.zip](https://github.com/trezor/trezord-go/files/1850280/signedWithSha256Timestamp.zip)
